File: output/deepseek_chat/avatar/algo_based_translation/python/java/atcoder_ABC145_E.java

import java.util.*;
import java.io.*;

public class atcoder_ABC145_E {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final int TIME = 0;
        final int VAL = 1;
        
        String[] firstLine = br.readLine().split(" ");
        int N = Integer.parseInt(firstLine[0]);
        int T = Integer.parseInt(firstLine[1]);
        
        int[][] timeValue = new int[N + 1][2];
        timeValue[0][TIME] = -1;
        timeValue[0][VAL] = -1;
        
        for (int i = 1; i <= N; i++) {
            String[] line = br.readLine().split(" ");
            timeValue[i][TIME] = Integer.parseInt(line[0]);
            timeValue[i][VAL] = Integer.parseInt(line[1]);
        }
        
        Arrays.sort(timeValue, 1, N + 1, (a, b) -> Integer.compare(a[VAL], b[VAL]));
        
        int[][] dp = new int[N + 1][T];
        for (int i = 0; i <= N; i++) {
            Arrays.fill(dp[i], -1);
        }
        
        for (int t = 0; t < T; t++) {
            dp[0][t] = 0;
        }
        
        for (int n = 1; n <= N; n++) {
            dp[n][0] = 0;
            for (int t = 1; t < T; t++) {
                if (timeValue[n][TIME] > t) {
                    dp[n][t] = dp[n - 1][t];
                } else {
                    dp[n][t] = Math.max(dp[n - 1][t], timeValue[n][VAL] + dp[n - 1][t - timeValue[n][TIME]]);
                }
            }
        }
        
        int valAcum = timeValue[N][VAL];
        int t = T - 1;
        int maxVal = valAcum + dp[N - 1][t];
        
        for (int n = N - 1; n >= 1; n--) {
            valAcum += timeValue[n][VAL];
            t -= timeValue[n + 1][TIME];
            if (t < 0) {
                break;
            } else {
                maxVal = Math.max(maxVal, valAcum + dp[n - 1][t]);
            }
        }
        
        System.out.println(maxVal);
    }
}

File: dataset/avatar/Python/Code/atcoder_ABC145_E.py
import sys
from operator import itemgetter
input = sys.stdin.readline
TIME = 0
VAL = 1
N , T = [ int ( a ) for a in input ( ).split ( ) ]
time_value = [ ( - 1 , - 1 ) ] + [ None ] * ( N )
for i in range ( 1 , N + 1 ) :
    time_value [ i ] = tuple ( int ( a ) for a in input ( ).split ( ) )
time_value.sort ( key = itemgetter ( VAL ) )
dp = [ [ - 1 ] * T for _ in range ( N + 1 ) ]
for t in range ( 0 , T ) :
    dp [ 0 ] [ t ] = 0
for n in range ( 1 , N + 1 ) :
    dp [ n ] [ 0 ] = 0
    for t in range ( 1 , T ) :
        if time_value [ n ] [ TIME ] > t :
            dp [ n ] [ t ] = dp [ n - 1 ] [ t ]
        else :
            dp [ n ] [ t ] = max ( dp [ n - 1 ] [ t ] , time_value [ n ] [ VAL ] + dp [ n - 1 ] [ t - time_value [ n ] [ TIME ] ] )
val_acum = time_value [ N ] [ VAL ]
t = T - 1
max_val = val_acum + dp [ N - 1 ] [ t ]
for n in range ( N - 1 , 0 , - 1 ) :
    val_acum += time_value [ n ] [ VAL ]
    t -= time_value [ n + 1 ] [ TIME ]
    if t < 0 :
        break
    else :
        max_val = max ( max_val , val_acum + dp [ n - 1 ] [ t ] )
print ( max_val )

